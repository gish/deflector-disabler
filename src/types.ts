export type SRFeed = {
  sr: {
    copyright: string;
    pagination: SRFeedPagination;
    episodes: { episode: SRFeedEpisode[] };
  };
};

type SRFeedPagination = {
  page: number;
  size: number;
  totalhits: number;
  totalpages: number;
  nextpage: string;
};

type SRFeedEpisode = {
  title: string;
  description: string;
  url: string;
  imageurl: string;
  downloadpodfile: {
    url: string;
    publishdateutc: string;
  };
};
