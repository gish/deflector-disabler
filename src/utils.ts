import { XMLParser } from "fast-xml-parser";
import { SRFeed, SRFeedEpisode } from "./types";
import { Podcast } from "podcast-rss";

export const generateFeed = (episodes: SRFeedEpisode[]): string => {
  const firstEpisode = episodes[0];
  const feed = new Podcast({
    title: "Late Spring in Program One",
    description:
      "Avsnitt av Sommar i P1 från 2024. Allt innehåll kommer från Sveriges Radio.",
    feedUrl: "",
    siteUrl: "",
    imageUrl: "https://gish.github.io/deflector-disabler/assets/cover.png",
    copyright: "Sveriges Radio",
    pubDate: firstEpisode
      ? firstEpisode.broadcast.playlist.publishdateutc
      : new Date(),
  });
  episodes.forEach((episode) => {
    feed.addItem({
      title: episode.title,
      description: episode.description,
      url: episode.url,
      guid: episode.downloadpodfile.url,
      date: episode.downloadpodfile.publishdateutc,
      enclosure: {
        url: episode.downloadpodfile.url,
      },
      itunesImage: episode.imageurl,
    });
  });
  return feed.buildXml();
};

export const parseFeed = (XMLData: string): SRFeed | null => {
  try {
    const parser = new XMLParser();
    return parser.parse(XMLData);
  } catch (_) {
    console.log("Failed parsing feed");
    return null;
  }
};
