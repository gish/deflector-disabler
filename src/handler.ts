import { existsSync, mkdirSync, writeFileSync } from "fs";
import { RadioProgram, SRFeedEpisode } from "./types";
import { DatabaseSync } from "node:sqlite";
import {
  getAllProgramEpisodes,
  getAllPrograms,
  saveProgramEpisodes,
  setProgramUpdatedTimestamp,
} from "./utils.database";
import { generatePodcastFeed, getFeed, parseFeed } from "./utils.feed";

function formatDate(timestamp: number): string {
  const d = new Date(timestamp);
  let month = "" + (d.getMonth() + 1);
  let day = "" + d.getDate();
  const year = d.getFullYear();

  if (month.length < 2) month = "0" + month;
  if (day.length < 2) day = "0" + day;

  return [year, month, day].join("-");
}

const writeFeedFile = (
  path: string,
  filename: string,
  podcastFeed: string,
): boolean => {
  if (!existsSync(path)) {
    mkdirSync(path);
  }
  try {
    writeFileSync(`${path}/${filename}.rss`, podcastFeed);
    return true;
  } catch (_) {
    return false;
  }
};

export const fetchProgramEpisodes = async (
  programId: number,
  lastRun: number,
  now: number,
): Promise<SRFeedEpisode[] | null> => {
  const ONE_DAY_MS = 24 * 60 * 60 * 1e3;
  /**
   * Get episodes published the day after last run until tomorrow.
   */
  const formattedFromDate = formatDate(lastRun + ONE_DAY_MS);
  const formattedToDate = formatDate(now + ONE_DAY_MS);
  const url = `https://api.sr.se/api/v2/episodes/index?programid=${programId}&fromdate=${formattedFromDate}&todate=${formattedToDate}&audioquality=hi`;

  return getProgramEpisodesByUrl(url);
};

const getProgramEpisodesByUrl = async (
  url: string,
): Promise<SRFeedEpisode[] | null> => {
  const feed = await getFeed(url);

  if (!feed) {
    return null;
  }

  const episode = feed.sr.episodes.episode;
  if (!episode) {
    return null;
  }
  const episodes = Array.isArray(episode) ? episode : [episode];

  const nextPage = feed.sr.pagination.nextpage;
  if (nextPage) {
    const nextFeedEpisodes = await getProgramEpisodesByUrl(nextPage);

    if (!nextFeedEpisodes) {
      console.error("no next feed episodes", { nextPage });
      return episodes;
    }
    return [...episodes, ...nextFeedEpisodes];
  }

  return episodes;
};

export const handler = async (
  now: Date,
  database: DatabaseSync,
  feedFilePath: string,
) => {
  let success = true;

  const programs = getAllPrograms(now, database);
  for (const program of programs) {
    const newEpisodes = await fetchProgramEpisodes(
      program.srId,
      program.lastUpdated,
      now.getTime(),
    );

    if (!newEpisodes) {
      setProgramUpdatedTimestamp(program, now, database);
      continue;
    }

    saveProgramEpisodes(program, newEpisodes, database);

    const episodes = getAllProgramEpisodes(program, database);
    const generatedFeed = generatePodcastFeed(program, episodes);
    const writeFeedFileResult = writeFeedFile(
      feedFilePath,
      program.slug,
      generatedFeed,
    );
    if (!writeFeedFileResult) {
      console.error("failed writing file", { title: program.title });
      success = false;
    }
    setProgramUpdatedTimestamp(program, now, database);
  }
  return success ? 0 : 1;
};
