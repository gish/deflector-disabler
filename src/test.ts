import { DatabaseSync } from "node:sqlite";
import { RadioProgram, SRFeedEpisode } from "./types";
import { mkdtempSync, readFileSync } from "node:fs";
import { handler } from "./handler";
import { tmpdir } from "node:os";
import { join } from "node:path";
import timemachine from "timemachine";
import { addPrograms, createTables } from "./utils.database";

describe("write new episodes to database", () => {
  const database = new DatabaseSync(":memory:");
  beforeAll(() => {
    timemachine.config({ dateString: "December 25, 2024 13:12:59" });
  });
  beforeEach(() => {
    const fixturePath = `./tests/fixtures/feed.twoItems.xml`;
    const fixture = readFileSync(fixturePath);
    global.fetch = jest.fn(() =>
      Promise.resolve({
        text: () => Promise.resolve(fixture),
        ok: true,
      }),
    ) as jest.Mock;

    createTables(database, true);
    addPrograms(
      [
        {
          id: 1,
          srId: 22,
          title: "program title",
          description: "program description",
          slug: "program",
          imageUrl: "https://image/",
          lastUpdated: 1,
        },
      ],
      database,
    );
  });

  afterEach(() => jest.restoreAllMocks());

  const filePath = mkdtempSync(join(tmpdir(), "deflector-disabler"));

  it.todo("should get paginated episodes");

  it("should write episodes to database", async () => {
    const now = new Date("2022-01-03");
    await handler(now, database, filePath);

    const episodesInDbStatement = database.prepare(
      "SELECT title, description FROM episodes",
    );
    const episodesInDb = episodesInDbStatement.all() as SRFeedEpisode[];
    expect(episodesInDb.length).toBe(2);
    expect(episodesInDb.at(0)?.title).toEqual("Hampus Nessvold");
    expect(episodesInDb.at(1)?.title).toEqual("Årets Sommarvärdar presenteras");
  });

  it("should update programs last update timestamp with latest episodes timestamp", async () => {
    const now = new Date("2022-01-04");
    await handler(now, database, filePath);
    const query = database.prepare(
      "SELECT lastUpdated FROM programs WHERE id = 1",
    );
    const result = query.get() as RadioProgram;
    expect(result).not.toBeFalsy();
    expect(result.lastUpdated).toEqual(now.getTime());
  });

  it("should fetch episodes with right dates", async () => {
    const today = new Date("2022-01-04");
    const fetch = global.fetch as jest.Mock;
    await handler(today, database, filePath);
    expect(fetch).toHaveBeenCalledWith(
      "https://api.sr.se/api/v2/episodes/index?programid=22&fromdate=1970-01-02&todate=2022-01-05&audioquality=hi",
    );

    const nextDay = new Date("2022-01-12");
    await handler(nextDay, database, filePath);
    expect(fetch).toHaveBeenCalledWith(
      "https://api.sr.se/api/v2/episodes/index?programid=22&fromdate=2022-01-05&todate=2022-01-13&audioquality=hi",
    );
  });

  it("should match snapshot of generated file", async () => {
    const now = new Date("2022-01-04");
    await handler(now, database, filePath);
    const generatedFile = readFileSync(`${filePath}/program.rss`, "utf8");
    expect(generatedFile).toMatchSnapshot();
  });

  it("should handle empty list of episodes", async () => {
    const fixturePath = `./tests/fixtures/empty.episodes.xml`;
    const fixture = readFileSync(fixturePath);
    global.fetch = jest.fn(() =>
      Promise.resolve({
        text: () => Promise.resolve(fixture),
        ok: true,
      }),
    ) as jest.Mock;
    const now = new Date("2022-01-04");
    // This should not throw
    await handler(now, database, filePath);
  });
});
