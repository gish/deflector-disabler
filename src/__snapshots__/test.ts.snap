// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`episode list generates episode list 1`] = `
[
  {
    "audiopreference": "broadcast",
    "audiopresentation": "music",
    "audiopriority": "aac",
    "broadcast": {
      "availablestoputc": "2024-07-22T04:55:00Z",
      "broadcastfiles": {
        "broadcastfile": {
          "duration": 4595,
          "publishdateutc": "2024-06-22T04:55:00Z",
          "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2413771)]",
          "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9370581-hi.m4a",
        },
      },
      "playlist": {
        "duration": 4595,
        "publishdateutc": "2024-06-22T04:55:00Z",
        "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2413771)]",
        "url": "https://api.sr.se/api/radio/radio.aspx?type=broadcast&id=9370581&codingformat=.m4a&metafile=asx&quality=hi",
      },
    },
    "broadcasttime": {
      "endtimeutc": "2024-06-22T12:20:00Z",
      "starttimeutc": "2024-06-22T11:00:00Z",
    },
    "channelid": 132,
    "description": "Skådespelaren om varför Försvarsmakten bör undvika att kalla in just honom om kriget kommer.",
    "downloadpodfile": {
      "availablefromutc": "2024-06-22T04:55:00Z",
      "description": "",
      "duration": 2856,
      "filesizeinbytes": 45718551,
      "program": "",
      "publishdateutc": "2024-06-22T11:00:00Z",
      "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2413771)]",
      "title": "Hampus Nessvold",
      "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9370585.mp3",
    },
    "imageurl": "https://static-cdn.sr.se/images/2071/53526fb9-9302-40e2-8b4e-3681f4c5b889.jpg?preset=api-default-square",
    "imageurltemplate": "https://static-cdn.sr.se/images/2071/53526fb9-9302-40e2-8b4e-3681f4c5b889.jpg",
    "listenpodfile": {
      "availablefromutc": "2024-06-22T04:55:00Z",
      "description": "",
      "duration": 2856,
      "filesizeinbytes": 45718551,
      "program": "",
      "publishdateutc": "2024-06-22T11:00:00Z",
      "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2413771)]",
      "title": "Hampus Nessvold",
      "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9370585.mp3",
    },
    "photographer": "Mattias Ahlm",
    "program": "",
    "publishdateutc": "2024-06-22T11:00:00Z",
    "title": "Hampus Nessvold",
    "url": "https://sverigesradio.se/avsnitt/2413771",
  },
  {
    "audiopreference": "broadcast",
    "audiopresentation": "music",
    "audiopriority": "aac",
    "broadcast": {
      "availablestoputc": "2024-07-11T12:00:00Z",
      "broadcastfiles": {
        "broadcastfile": {
          "duration": 3600,
          "publishdateutc": "2024-06-11T11:00:00Z",
          "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2393740)]",
          "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9360224-hi.m4a",
        },
      },
      "playlist": {
        "duration": 3600,
        "publishdateutc": "2024-06-11T11:00:00Z",
        "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2393740)]",
        "url": "https://api.sr.se/api/radio/radio.aspx?type=broadcast&id=9360224&codingformat=.m4a&metafile=asx&quality=hi",
      },
    },
    "broadcasttime": {
      "endtimeutc": "2024-06-11T12:00:00Z",
      "starttimeutc": "2024-06-11T11:00:00Z",
    },
    "channelid": 132,
    "description": "Programchef Bibi Rödöö presenterar årets 58 Sommarvärdar.",
    "downloadpodfile": {
      "availablefromutc": "2024-06-11T12:00:00Z",
      "description": "",
      "duration": 3585,
      "filesizeinbytes": 57369891,
      "program": "",
      "publishdateutc": "2024-06-11T11:00:00Z",
      "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2393740)]",
      "title": "Årets Sommarvärdar presenteras",
      "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9360211.mp3",
    },
    "imageurl": "https://static-cdn.sr.se/images/2071/b4a8f917-2507-4b34-b14a-10d2731821f5.jpg?preset=api-default-square",
    "imageurltemplate": "https://static-cdn.sr.se/images/2071/b4a8f917-2507-4b34-b14a-10d2731821f5.jpg",
    "listenpodfile": {
      "availablefromutc": "2024-06-11T12:00:00Z",
      "description": "",
      "duration": 3585,
      "filesizeinbytes": 57369891,
      "program": "",
      "publishdateutc": "2024-06-11T11:00:00Z",
      "statkey": "/app/avsnitt/p1[k(132)]/sommar & vinter i p1[p(2071)]/[e(2393740)]",
      "title": "Årets Sommarvärdar presenteras",
      "url": "https://sverigesradio.se/topsy/ljudfil/srapi/9360211.mp3",
    },
    "program": "",
    "publishdateutc": "2024-06-11T11:00:00Z",
    "title": "Årets Sommarvärdar presenteras",
    "url": "https://sverigesradio.se/avsnitt/2393740",
  },
]
`;

exports[`parse and generate parses and generates feed with two items correctly 1`] = `"<?xml version="1.0" encoding="UTF-8"?><rss xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:atom="http://www.w3.org/2005/Atom" version="2.0" xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd"><channel><title><![CDATA[Late Spring in Program One]]></title><description><![CDATA[Avsnitt av Sommar i P1 från 2024. Allt innehåll kommer från Sveriges Radio.]]></description><link>http://github.com/dylang/node-rss</link><generator>Podcast for Node</generator><lastBuildDate>Sat, 01 Jun 2024 00:00:00 GMT</lastBuildDate><pubDate>Sat, 22 Jun 2024 04:55:00 GMT</pubDate><copyright><![CDATA[Sveriges Radio]]></copyright><itunes:summary>Avsnitt av Sommar i P1 från 2024. Allt innehåll kommer från Sveriges Radio.</itunes:summary><itunes:explicit>no</itunes:explicit><itunes:image href="https://gish.github.io/deflector-disabler/assets/cover.png"/><item><title><![CDATA[Hampus Nessvold]]></title><description><![CDATA[Skådespelaren om varför Försvarsmakten bör undvika att kalla in just honom om kriget kommer.]]></description><link>https://sverigesradio.se/avsnitt/2413771</link><guid isPermaLink="false">https://sverigesradio.se/topsy/ljudfil/srapi/9370585.mp3</guid><pubDate>Sat, 22 Jun 2024 11:00:00 GMT</pubDate><enclosure url="https://sverigesradio.se/topsy/ljudfil/srapi/9370585.mp3" length="0" type="audio/mpeg"/><itunes:summary>Skådespelaren om varför Försvarsmakten bör undvika att kalla in just honom om kriget kommer.</itunes:summary><itunes:explicit>no</itunes:explicit><itunes:image href="https://static-cdn.sr.se/images/2071/53526fb9-9302-40e2-8b4e-3681f4c5b889.jpg?preset=api-default-square"/></item><item><title><![CDATA[Årets Sommarvärdar presenteras]]></title><description><![CDATA[Programchef Bibi Rödöö presenterar årets 58 Sommarvärdar.]]></description><link>https://sverigesradio.se/avsnitt/2393740</link><guid isPermaLink="false">https://sverigesradio.se/topsy/ljudfil/srapi/9360211.mp3</guid><pubDate>Tue, 11 Jun 2024 11:00:00 GMT</pubDate><enclosure url="https://sverigesradio.se/topsy/ljudfil/srapi/9360211.mp3" length="0" type="audio/mpeg"/><itunes:summary>Programchef Bibi Rödöö presenterar årets 58 Sommarvärdar.</itunes:summary><itunes:explicit>no</itunes:explicit><itunes:image href="https://static-cdn.sr.se/images/2071/b4a8f917-2507-4b34-b14a-10d2731821f5.jpg?preset=api-default-square"/></item></channel></rss>"`;
